03_Modules/Locations/src/module/admin.ts

```diff
@@ -0,0 +1,172 @@
+import { Service } from 'typedi';
+import { type ILogObj, Logger } from 'tslog';
+import {
+  ChargingStation,
+  Location,
+  SequelizeLocationRepository,
+} from '@citrineos/data';
+import {
+  AdminConnectorDTO,
+  AdminEVSEDTO,
+  AdminLocationDTO,
+  CitrineOcpiLocationMapper,
+  LocationDTO,
+  LocationsBroadcaster,
+  OcpiLocation,
+  OcpiLocationProps,
+  OcpiLocationRepository,
+  OcpiEvse,
+  OcpiEvseRepository,
+  OcpiConnector,
+  OcpiConnectorRepository,
+  VariableAttributesUtil,
+  OcpiLocationsUtil,
+  InvalidParamException,
+} from '@citrineos/ocpi-base';
+
+@Service()
+export class AdminLocationsService {
+  constructor(
+    private logger: Logger<ILogObj>,
+    private locationMapper: CitrineOcpiLocationMapper,
+    private locationRepository: SequelizeLocationRepository,
+    private ocpiLocationRepository: OcpiLocationRepository,
+    private ocpiEvseRepository: OcpiEvseRepository,
+    private ocpiConnectorRepository: OcpiConnectorRepository,
+    private locationsBroadcaster: LocationsBroadcaster,
+    private variableAttributesUtil: VariableAttributesUtil,
+    private ocpiLocationsUtil: OcpiLocationsUtil,
+  ) {}
+
+  public async createOrUpdateLocation(
+    adminLocationDto: AdminLocationDTO,
+    broadcast: boolean,
+  ): Promise<LocationDTO> {
+    this.logger.debug(
+      adminLocationDto.id
+        ? `Updating Location ${adminLocationDto.id}`
+        : `Creating Location ${adminLocationDto.name}`,
+    );
+
+    const [invalidLocation, message] = AdminLocationDTO.IS_LOCATION_INVALID(adminLocationDto);
+    if (invalidLocation) {
+      throw new InvalidParamException(message ?? 'Invalid location request.');
+    }
+
+    const [ocpiLocation, citrineLocation] =
+      this.mapAdminLocationDtoToEntities(adminLocationDto);
+    const savedCitrineLocation =
+      await this.locationRepository.createOrUpdateLocationWithChargingStations(
```

Comment:
Could this method `createOrUpdateLocationWithChargingStations` not just be named `createOrUpdateLocation`?

Reply:
i only made this explicit in the event that someone just wants to update locations without charging stations involved (given that that's how most of our `createOrUpdate` methods are named and function, they don't necessarily touch the nested objects), but if that doesn't come to pass then we can update back to createOrUpdateLocation